### PR TITLE
Improve metaclass detection in nested scopes in Python 3

### LINKED ISSUE #1177: Spurious unused-variable warning when function-scoped class is used only as metaclass
### Code that reproduces the error:
```
def func_scope():
    class Meta(type):
        pass
    class Class(metaclass=Meta):
        pass
    return Class
```

### Current behavior
Generates a warning: `W:  2, 4: Unused variable 'Meta' (unused-variable)`
(And various valid complaints, mostly about docstrings.)

### Expected behavior
The warning should not be generated.

### pylint --version output
pylint 1.6.4,
astroid 1.4.8
Python 3.5.2 (default, Oct 11 2016, 05:05:28)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.38)]

I have a probably-stale checkout of pylint from a previous bug I reported. I ran it against the same file, and got the same output.

pylint 2.0.0,
astroid 1.5.0
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  5 2015, 21:12:44)
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]

### PR DESCRIPTION
**Pull request by rogalski**
TODO:

- [x] write a test that exposes an issue
- [x] implement changes to fix exposed issue
- [x] improve coverage by adding tests for potential failures
- [x] make all new tests pass
- [x] refactor code
- [x] add docs and change log entries

Current fail is triggered, when metaclass is imported in non local (enclosing) scope. I’m not quite sure yet how pushing and popping from `_to_consume` stack should work, but second element of pushed tuple seems to be never populated.

Any input is appreciated.

Closes #1177

### CODE REVIEW

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -1203,35 +1203,43 @@ def leave_listcomp(self, _):
...
-        """ Update consumption analysis variable
-        for metaclasses.
-        """
-        module_locals = self._to_consume[0][0]
-        module_imports = self._to_consume[0][1]
-        consumed = {}
+        self._check_metaclasses(node)
+        super(VariablesChecker3k, self).leave_module(node)
+
+    def _check_metaclasses(self, node):
```
Comment (rogalski) [inline pylint/checkers/variables.py]: I feel like naming of variables  `consumed `, `scope_not_consumed_locals`, `scope_locals_dict` should be better.

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -1203,35 +1203,43 @@ def leave_listcomp(self, _):
...
+        """ Update consumption analysis for metaclasses. """
+        consumed = []  # (scope_locals_dict, key)
+
+        for klass in node.get_children():
+            if not isinstance(klass, astroid.ClassDef):
```
Comment (rogalski) [inline pylint/checkers/variables.py]: Maybe this loop body should be extracted to method.

File: pylint/checkers/variables.py [hunk 3]
```diff
@@ -1203,35 +1203,43 @@ def leave_listcomp(self, _):
...
-            # and in the imported modules.
+            name = None
             if isinstance(klass._metaclass, astroid.Name):
                 name = klass._metaclass.name
             elif metaclass:
-                # if it uses a `metaclass=module.Class`
                 name = metaclass.root().name
 
+            found = None
             if name:
```
Comment (rogalski) [inline pylint/checkers/variables.py]: Maybe if body should be extracted to method.

File: pylint/checkers/variables.py [hunk 4]
```diff
@@ -1203,35 +1203,43 @@ def leave_listcomp(self, _):
...
-                found = consumed.setdefault(
-                    name, module_locals.get(name, module_imports.get(name)))
+                # check enclosing scopes starting from most local
```
Comment (rogalski) [inline pylint/checkers/variables.py]: Maybe if body should be extracted to method.

Comment (PCManticore) [inline pylint/checkers/variables.py]: Yes, let's do that.

File: pylint/checkers/variables.py [hunk 5]
```diff
@@ -1203,35 +1203,44 @@ def leave_listcomp(self, _):
...
-        consumed = {}
+        self._check_metaclasses(node)
+        super(VariablesChecker3k, self).leave_module(node)
+
+    def _check_metaclasses(self, node):
+        """ Update consumption analysis for metaclasses. """
+        consumed = []  # [(scope_locals, consumed_key)]
+
+        for child_node in node.get_children():
+            if not isinstance(child_node, astroid.ClassDef):
```
Comment (rogalski) [inline pylint/checkers/variables.py]: I want to check just this scope, not do full recursive search.

**Comment by PCManticore** [comment]
Thank you!

**Pull request merged**